### PR TITLE
Exercise 2

### PR DESCRIPTION
**Pull request by CreaTiVeUK**
Please make sure you are raising this PR against your own repository and not the original. If it says https://github.com/CorndelWithSoftwire in the search bar right now, you are in the wrong place!

Please also don't forget to submit the PR's url to Aptem afterwards (by clicking on the corresponding exercise component, pasting the url in the textbox that appears and then pressing the finish button).

### CODE REVIEW

File: todo_app/data/item.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        """
+            Create a card as part of a list with the specified list name and card name.
+
+        Args:
+            list_name: List name.
+            card_name: Card name.
+        """
+        card = {
+            "name": card_name,
+            "idBoard": "61c9efb12a4989214a75b4a5",
```
Comment (hugh-emerson) [inline todo_app/data/item.py:26]: Use the `board_id` variable that was loaded from an env var.

File: todo_app/data/item.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
+from todo_app.config.credentials import *
```
Comment (hugh-emerson) [inline todo_app/data/item.py:1]: From the [PEP 8 style guide](https://peps.python.org/pep-0008/):

> Wildcard imports (from <module> import *) should be avoided, as they make it unclear which names are present in the namespace, confusing both readers and many automated tools

Basically, it's not clear when reading the code that `api_token` came from the credentials module, and you might accidentally define new variables here that reuse the names etc.

File: todo_app/data/session_items_trello.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    response = requests.get(url)
+    response_json = json.loads(response.text)
+    cards, lists = [], []
+    for list in response_json:
+        lists.append(List(list['id'], list['name']))
+        for card in list['cards']:
+            if card['due'] is not None:
+                due = dateutil.parser.parse(card['due']).strftime('%m/%d/%Y')
+            else:
+                due = card['due']
```
Comment (hugh-emerson) [inline todo_app/data/session_items_trello.py:26]: Confusing. Isn't this equivalent to `due = None`? Given the if/else.

File: todo_app/logs/app.log [hunk 4]
```diff
@@ -0,0 +1,159 @@
+### 2022-02-27 18:52:11,808 ### appLog ### INFO ### Title to add: rweqrwer ###
```
Comment (hugh-emerson) [inline todo_app/logs/app.log:1]: Don't commit log files. 

Add it to your .gitignore file and remove it from the repo (either delete it altogether and commit, or run `git rm --cached todo_app/logs/app.log`

File: todo_app/templates/layout.html [hunk 5]
```diff
@@ -1,24 +1,40 @@
...
+  <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.7/umd/popper.min.js"
+    integrity="sha384-UO2eT0CpHqdSJQ6hJty5KVphtPhzWj9WO1clHTMGa3JDZwrnQq4sF86dIHNDz0W1"
+    crossorigin="anonymous"></script>
+  <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/js/bootstrap.min.js"
+    integrity="sha384-JjSmVgyd0p3pXB1rRibZUAYoIIy6OrQ6VrjIEaFf/nJGzIxFDsf4x0xIM+B07jRM"
+    crossorigin="anonymous"></script>
+  <script>
+    $(function () {
+      $(".date").datepicker({ dateFormat: 'mm/dd/yy' }).val();
+    });
```
Comment (hugh-emerson) [inline todo_app/templates/layout.html:36]: Note you could have tried using `<input type="date">` for the browser's built-in datepicker. But the one you've added supports more customisation etc.

File: README.md [hunk 6]
```diff
@@ -32,9 +33,11 @@ $ cp .env.template .env  # (first time only)
 
 The `.env` file is used by flask to set environment variables when running `flask run`. This enables things like development mode (which also enables features like hot reloading when you make a file change). There's also a [SECRET_KEY](https://flask.palletsprojects.com/en/1.1.x/config/#SECRET_KEY) variable which is used to encrypt the flask session cookie.
 
+You'll need to sign up to trello and insert your Key & Token in the .env file.
```
Comment (hugh-emerson) [inline README.md:36]: You haven't included useful details. The reader has to go and read the codebase to work out what the env vars should be called. And you also haven't mentioned the fact that they need `TRELLO_URL` and `TRELLO_BOARD_ID` as well.

Nicer would be to add all 4 variables with placeholder values to .env.template. Then it's easier for the reader to get set up.

**Review by hugh-emerson — APPROVED** [review]
Looks good! I've just left some optional suggestions for things you could tidy up.

Merge when you're ready 👍

File: README.md [hunk 7]
```diff
@@ -22,6 +22,7 @@ The project uses a virtual environment to isolate package dependencies. To creat
 
 ```bash
 $ poetry install
+$ pip install python-dateutil
```
[the remark this answers is not in the document]
Comment (CreaTiVeUK) [reply README.md:25]: Thanks for the tip - pyproject.toml updated and removed the comment from README file.

File: todo_app/data/item.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+    def __init__(self, id, name, description, due, list_name, list_id):
+        self.id = id
+        self.name = name
+        self.description = description
+        self.due = due
+        self.list_name = list_name
+        self.list_id = list_id
+
+    def create_Card(list_name, card_name):
```
[the remark this answers is not in the document]
Comment (CreaTiVeUK) [reply todo_app/data/item.py:16]: You are right. Fixed now.

File: todo_app/config/credentials.py [hunk 9]
```diff
@@ -0,0 +1,9 @@
+import os
+import dotenv
+
+dotenv.load_dotenv()
```
[the remark this answers is not in the document]
Comment (CreaTiVeUK) [reply todo_app/config/credentials.py:4]: Noted. Resolved

File: todo_app/data/item.py (hunk 1 above)
Comment (CreaTiVeUK) [reply todo_app/data/item.py:26]: Fixed

File: todo_app/data/item.py (hunk 2 above)
Comment (CreaTiVeUK) [reply todo_app/data/item.py:1]: Fixed

File: todo_app/data/session_items_trello.py (hunk 3 above)
Comment (CreaTiVeUK) [reply todo_app/data/session_items_trello.py:26]: Correct. 
-> If date's provided, reformat and assign to due.
-> Else, assign None to due.

File: todo_app/logs/app.log (hunk 4 above)
Comment (CreaTiVeUK) [reply todo_app/logs/app.log:1]: Added to .gitignore on top level and removed log.

File: todo_app/templates/layout.html (hunk 5 above)
Comment (CreaTiVeUK) [reply todo_app/templates/layout.html:36]: Main reason is it complies with the American date format Trello uses.

**Pull request merged**